client/lib/protect-form/index.jsx

```diff
@@ -15,8 +15,37 @@ const confirmText = i18n.translate( 'You have unsaved changes. Are you sure you
 const beforeUnloadText = i18n.translate( 'You have unsaved changes.' );
 let formsChanged = [];
 
+function warnIfChanged( event ) {
+	if ( ! formsChanged.length ) {
+		return;
+	}
+	debug( 'unsaved form changes detected' );
+	( event || window.event ).returnValue = beforeUnloadText;
+	return beforeUnloadText;
+}
+
+function markChanged( form ) {
+	if ( ! includes( formsChanged, form ) ) {
+		formsChanged.push( form );
+	}
+}
+
+function markSaved( form ) {
+	formsChanged = without( formsChanged, form );
+}
+
+/*
+ * HOC that passes markChanged/markSaved props to the wrapped component instance
+ */
 export const protectForm = WrappedComponent => {
 	return class ProtectedFormComponent extends Component {
+		// This function binding makes the listener attached to window unique for each
+		// component instance.
+		warnIfChanged = warnIfChanged.bind();
+
+		markChanged = markChanged.bind( null, this );
+		markSaved = markSaved.bind( null, this );
```

Comment:
this is an odd construction the should probably have an explanation. usually one would expect `this` to be in the first argument but here it's being explicitly set so that the `this` in `markChanged` and `markSaved` is `null`/global but the argument `form` is equal to the `this` here at the bind-site.

maybe this is evidence that we could rewrite these functions to simplify things a bit. did we consider using arrow-functions to preserve lexical binding of `this`?